### PR TITLE
Typescript

### PR DESCRIPTION
**Pull request by 12wrigja**
This PR begins the process of migrating the Temporal polyfill's sources to TypeScript.

I'm less familiar with the impact of this change on the distributed NPM bundle / changes to package.json and package-lock.json, so a review from someone with that expertise would be greatly appreciated.

In follow-up PRs I will migrate various parts of the source to more canonical representations (e.g. Calendars should probably be based on an abstract class hierarchy).

### CODE REVIEW

File: lib/ecmascript.ts
Comment (shicks) [inline lib/ecmascript.ts:124]: `ReadonlyArray<readonly [string, string]>` ?

What about just adding `as const` to the end of the array literal?

File: lib/init.js
Comment (shicks) [inline lib/init.js:6]: Do we need Closure Compiler to be able to handle this?  Does it?

Comment (12wrigja) [reply lib/init.js:6]: We don't - this file is only used to launch the polyfill Playground.

File: lib/calendar.ts
Comment (ptomato) [inline lib/calendar.ts:362]: Maybe @justingrant can review this part?

File: tsconfig.json
Comment (ptomato) [inline tsconfig.json:42]: This is probably my unfamiliarity with TypeScript configuration, but how is it outputting ESM files for an ES5 target?

File: rollup.config.js
Comment (12wrigja) [reply rollup.config.js:17]: Yes, because this replacement happens over the emitted js, not the source ts file.

I tested this by running `NODE_ENV=production npm run build`, and grepping the output file for `_repr_`, which is one of the common properties that is only set in debug builds - it is still removed from the build with this change.

File: lib/calendar.ts
Comment (12wrigja) [reply lib/calendar.ts:1094]: The error is 

```
Conversion of type <omitted for brevity> to type 'NonIsoHelperBase' may be a mistake because neither type sufficiently overlaps with the other. If this was intentional, convert the expression to 'unknown' first.
```

I believe they don't overlap because the `this` in that context is being inferred as the collection of properties only within that object literal (and doesn't include all the properties that are present due to the `nonIsoHelperBase` within the Object.assign call). This seems to be a limitation of TypeScript's type inference, but I think it could be resolved by rewriting the calendars using classes (where this sort of extension is better understood in the type system).

File: tsconfig.json
Comment (12wrigja) [reply tsconfig.json:42]: https://github.com/Microsoft/TypeScript/issues/6319
https://github.com/microsoft/TypeScript/issues/5790

These imply to me that the options here are asking tsc to emit ES5 code using the ES6 module format instead of defaulting to commonjs. It seems the module format is only loosely tied to the language output level.

I too don't know for sure if this is "right" - I was basing it off the internal tsconfig used at Google but with modifications to get the existing test suite to run (which needed ESM). If we were to change a value here, I think it would be to change target to something higher and let the rest of the rollup config handle that output.

File: lib/ecmascript.ts
Comment (12wrigja) [reply lib/ecmascript.ts:124]: Done, though this required a small adjustment to the implementation of ES.DefaultTemporalLargestUnit to compile again (though I think this change would be needed anyways as the existing implementation was not renaming-safe for Closure Compiler).

File: tsconfig.json
Comment (12wrigja) [reply tsconfig.json:42]: That being said, I did adjust the module output to explicitly be es6 instead of the unstable esnext.

**Comment by justingrant** [comment]
I'll review this PR tomorrow. One immediate suggestion is to make sure to rebase on top of the most recent polyfill code from proposal-temporal because a few minor PRs have landed there recently. Also, are there any changes that you made that should be back-ported to the old MJS polyfill?

**Comment by 12wrigja** [comment]
I opened https://github.com/js-temporal/temporal-polyfill/pull/22 to port over the normative changes I saw in the proposal repo, though I'm not entirely sure how they will translate over to TS (unsure how you'd assign those properties to what are currently module exports). Once that is merged then I'll rebase this.

Regarding backporting: there's a fair number of changes that could be backported, but I'd prefer to wait for this to be merged before starting that. I'd probably consider backporting:
-  a fair number of let->const changes that the linter decided it wanted fixing (though why it didn't seem to trigger on the existing mjs source is confusing to me).
 - I adjusted how the `__debug__` worked to make tsc happy (it doesn't understand the global comments, and I'd prefer not to pollute global typings by re-opening the window/global namespace to include that symbol).
 - various copies of the `bigIntIfAvailable` function needed to check for `BigInt` off of `globalThis` (instead of doing so implicitly)
 - renaming the default timezone function (this seems to greatly confuse tsc, but was clearly valid code before)

Everything else are changes that are in TypeScript type space (or are adjustments to use tsc modules instead of Node mjs), so I don't think can be backported unless we moved that impl to TS source as well (which could be done if people felt it was useful?).

**Comment by ptomato** [comment]
I'm afraid backporting anything that wasn't clearly a bug fix to the old impl would end up being contentious within TC39.

**Comment by 12wrigja** [comment]
Ah right. In that case, I don't think there are any bug-fix-like changes to backport.

File: lib/now.ts
Comment (justingrant) [inline lib/now.ts:8]: Out of curiosity, why was this renaming needed?

File: lib/calendar.ts
Comment (justingrant) [reply lib/calendar.ts:362]: Yeah, this looks like a bug. Probably a holdover from a previous merge-based implementation.

File: lib/calendar.ts
Comment (justingrant) [inline lib/calendar.ts:1151]: Instead of these casts, would it make sense to put a `this: NonIsoHelperBase` into the param declaration like you did in other files?

File: lib/calendar.ts
Comment (justingrant) [inline lib/calendar.ts:1855]: Would it make sense to abstract this type into a global type declaration to avoid repeating `typeof nonIsoGeneralImpl & { helper: NonIsoHelperBase }` in lot of places?

**Review by justingrant — APPROVED** [review]
Overall looks good. Thanks! I made a bunch of minor suggestions and asked a few questions. 

The `${dt}` thing (see my comment) is concerning though... if that really breaks in all TS code, then we may want to discuss changing `valueOf` behavior to avoid throwing every time a TS user tries to console.log a template string with a Temporal instance!

File: rollup.config.js
Comment (justingrant) [reply rollup.config.js:17]: While you're changing this, could you also fix the rollup warning with this plugin?  Here's what I see when I run `npm build` on this branch:
```
(!) Plugin replace: @rollup/plugin-replace: 'preventAssignment' currently defaults to false. It is recommended to set this option to `true`, as the next major version will default this option to `true`.
```

Comment (justingrant) [reply rollup.config.js:17]: Er, never mind the comment above-- I was on the wrong branch.  Build works fine.

**Review by justingrant — CHANGES_REQUESTED** [review]
I changed my mind about approval. ;-)

There's a problem with sourcemaps that will make it impossible to set breakpoints or step into this polyfill's original source using a the debugger of browser dev tools or an IDE debugger like VS Code. The problem is that the output sourcemap has incorrect relative paths for Temporal's own source files. 

If you look at /dist/index.js.map, you'll see this at the end of the `files` array: 

```
"../../lib/intrinsicclass.ts","../../lib/slots.ts","../../lib/calendar.ts","../../lib/regex.ts","../../lib/ecmascript.ts","../../lib/timezone.ts","../../lib/intl.ts","../../lib/instant.ts","../../lib/plaindate.ts","../../lib/plaindatetime.ts","../../lib/duration.ts","../../lib/plainmonthday.ts","../../lib/now.ts","../../lib/plaintime.ts","../../lib/plainyearmonth.ts","../../lib/zoneddatetime.ts","../../lib/legacydate.ts"
```

These have one too many `../` which will make the sources unfindable when installed from NPM. 

instead, the `files` list should have paths like this: `"../lib/intrinsicclass.ts"` which reflects the correct relative path between `lib` and `dist` folders. 

The root cause is this line in tsconfig.json:

```json
 "outDir": "dist/esm",
```

The rollup typescript plugin is presumably building a sourcemap assuming that output files will be living in `dist/esm` when in reality they're living in `dist`.  Simply changing tsconfig.json to "outDir": "dist" will break tests, so there's some research req'd to figure this out.

**Comment by justingrant** [comment]
Here's another, more minor sourcemap isssue: I use a webpack plugin to detect differences between code stored inline in the  sourcemap vs original source files. My plugin complained about the two lines below. Any idea what's going on here?  

```
/Users/justingrant/Documents/hdev/h3/web/node_modules/tslib/tslib.es6.js doesn't match sourcemap, starting at line 20, col 48
Original Line 19:         ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
     Map Line 19:         ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
Original Line 20:         function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
     Map Line 20:         function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
Original Line 21:     return extendStatics(d, b);
     Map Line 21:     return extendStatics(d, b);

/Users/justingrant/Documents/hdev/h3/web/node_modules/is-date-object/index.js doesn't match sourcemap, starting at line 15, col 54
Original Line 14: var dateClass = '[object Date]';
     Map Line 14: var dateClass = '[object Date]';
Original Line 15: var hasToStringTag = typeof Symbol === 'function' && !!Symbol.toStringTag;
     Map Line 15: var hasToStringTag = typeof Symbol === 'function' && typeof Symbol.toStringTag === 'symbol';
Original Line 16: 
     Map Line 16: 
```

**Comment by justingrant** [comment]
A final sourcemap problem is related to the fact that our polyfill when installed in a project will be 2 levels under node_modules (first `@js-temporal` and seond `polyfill`).  So either the sourcemap should be rewritten to add an extra `../` for all references to node_modules code, or we could try rewriting the path to `~/` in place of `../node_modules`.  The `~/` is AFAICT an undocumented synonym for "the app's node_modules folder", although I don't know if it has universal support. See https://github.com/justingrant/amplify-js/blob/768ad63b2689d377eddafe3802c50a31f509e5ab/packages/aws-amplify/webpack-utils.js#L56-L60 for an example of this rewriting that I did an an AWS PR a while ago. 

The other line of investigation on this issue is why we're including code in our sourcemaps from our dependency libraries. Does this mean that rollup and/or TS includes es-abstract, big-integer, and their transitive dependencies in our bundle?  That's probably not what we want. Is there a rollup and/or TS setting to prevent that?

**Comment by justingrant** [comment]
> Does this mean that rollup and/or TS includes es-abstract, big-integer, and their transitive dependencies in our bundle?

BTW, here's a relatively modern TS library and the rollup config they use: https://github.com/mongodb/js-bson/blob/master/rollup.config.js.

If you `npm i bson` and look at the sourcemaps (and likely the bundled JS files too) you'll see that bson.bundle.js.map contains `"../node_modules/buffer/index.js" (it bundles its dependencies) but bson.esm.js.map does not bundle dependencies. The latter output file comes from a rollup config with `format: 'es'` that also uses the `external` config option which presumably excludes dependencies.

Would a similar ESM config be worth adding to this polyfill?

File: lib/ecmascript.ts
Comment (12wrigja) [reply lib/ecmascript.ts:1329]: `throw new RangeError(`Offset ${offsetStr} is invalid for ${dt} in ${timeZoneString}`);` is transpiled by tsc to `throw new RangeError('Offset ' + offsetStr + ' is invalid for ' + dt + ' in ' + timeZoneString);`, and the '+' operator invokes `valueOf`.

This would be an issue for either language given that the problem occurs regardless of who is using tsc to compile the sources - us when we bundle the sources for packaging or others I suppose if they directly compile against our TS source.

Is having `valueOf` throwing something officially part of the Temporal spec, and thus something that the native object will do as well?

File: lib/ecmascript.ts
Comment (12wrigja) [reply lib/ecmascript.ts:373]: It isn't a number, but it is typed as such to "cheat" and allow that variable to be reassigned to a number below (presumably to save on memory allocation?).

Naively, I would have pulled these out in `const`s and then create a new variable to store the numeric conversion of this value, but I was attempting to avoid actual code changes (vs just type-space additions) so the runtime behaviour wouldn't change very much.

File: lib/ecmascript.ts
Comment (justingrant) [reply lib/ecmascript.ts:1329]: > Is having `valueOf` throwing something officially part of the Temporal spec, and thus something that the native object will do as well?

Yes. https://tc39.es/proposal-temporal/#sec-temporal.plaindate.prototype.valueof. That's why I'm concerned about this.

I'll add this issue to the agenda for the next Temporal Champions meeting. We may want to keep the behavior as-is, but transpiling string interpolation to `+` might be popular enough (perhaps not only in TS, maybe in Babel too for regular JS?) that we may want to revisit.

Comment (justingrant) [reply lib/ecmascript.ts:1329]: To clarify, this isn't an issue for this PR-- it's a more general issue for Temporal overall. See https://github.com/tc39/proposal-temporal/issues/1681. Feel free to resolve this conversation.

File: lib/ecmascript.ts
Comment (justingrant) [reply lib/ecmascript.ts:373]: Sounds good, thanks.

File: lib/now.ts
Comment (12wrigja) [reply lib/now.ts:8]:
```
Parameter 'temporalTimeZoneLike' cannot reference identifier 'timeZone' declared after it.ts(2373)
Block-scoped variable 'timeZone' used before its declaration.ts(2448)
now.ts(9, 9): 'timeZone' is declared here.
```

It seems that tsc manages to confuse the timeZone function and the constant declared within the function body, so I renamed the function.

File: lib/calendar.ts
Comment (12wrigja) [reply lib/calendar.ts:1855]: When I tried this I got an error: `Type alias 'NonIsoGeneralImplThis' circularly references itself.ts(2456)`.

Example:
```
type NonIsoGeneralImplThis = typeof nonIsoGeneralImpl & { helper: NonIsoHelperBase };

... 

const nonIsoGeneralImpl = {
  dateFromFields(this: NonIsoGeneralImplThis, fields, options, calendar) {
    const overflow = ES.ToTemporalOverflow(options);
```

File: lib/ecmascript.ts
Comment (12wrigja) [reply lib/ecmascript.ts:1329]: Thanks for opening that issue. I've added a reference to the TypeScript repo issue so if someone stumbles across this fix in the future and finds out it's been fixed upstream we can revert this fix.

File: lib/ecmascript.ts
Comment (12wrigja) [reply lib/ecmascript.ts:1918]: Done.

File: lib/calendar.ts
Comment (12wrigja) [reply lib/calendar.ts:1151]: Done - replaced all the `as unknown as` with the equivalent `this` type.

File: lib/calendar.ts
Comment (12wrigja) [reply lib/calendar.ts]: I swapped this over to specifying the `this` type as part of the function's parameters to avoid the dangerous casts. Resolving.

**Comment by 12wrigja** [comment]
I've pushed another commit that I believe addresses your comments - it's heavily based on the commit you sent me with some small adjustments. In particular, we don't need to have two tsconfig files as the rollup TypeScript plugin can use the package tsconfig as a base and seems to adjust various settings automatically to work with the rest of the build tools.

I'll wait until #25 is merged before rebasing and squashing / cleaning up the commits.

File: lib/calendar.ts
Comment (12wrigja) [reply lib/calendar.ts:362]: In that case it should probably get backported right? If I understand the code right, in a hypothetical scenario where one was copying a "full cache" (has MAX_CACHE_ENTRIES things in it), there would be nothing in the new cache?

Comment (justingrant) [reply lib/calendar.ts:362]: The consequence of this is purely performance, not functional, so unless we're seeing aberrant perf caused by cache exhaustion (AFAIK we're not) then there's no need to back-port to the old polyfill that's just there for running tests. AFAIK there are no tests that fill up the cache, BTW.

File: tsconfig.json
Comment (justingrant) [inline tsconfig.json:41]: Changes look great. Sourcemaps and tests both work. Minor suggestion: the folder name `ts-dist` is too close to "dist" for my taste, given that the former is only for internal use while the latter is published to npm. Possible to pick another name with "test" or "internal" in the name and "dist" not in it?

Comment (12wrigja) [reply tsconfig.json:41]: I've moved this to ./tsc-out/ so it's very clear what is putting files there.

File: lib/now.ts
Comment (justingrant) [reply lib/now.ts:8]: In this case, I'd suggest renaming the local variables `tz` instead of renaming the `timeZone` function. This would be less change vs. JS, and also IMHO it'd be more clear that the default params are calling the implementation of `Now.timeZone` as opposed to some other function.

Comment (12wrigja) [reply lib/now.ts]: Done - renamed as you suggested.

File: lib/calendar.ts
Comment (justingrant) [reply lib/calendar.ts:1870]: Ahhh, OK. Good thing for a future black-belt TS contributor to figure out. Not req'd now.

BTW, I apologize for the TS-unfriendly structure of this code. I wrote it this way to simplify adapting the code from the previous non-ISO-calendar implementation which used a similar class-like-but-not-really structure for these singletons. If I were starting all over again I'd probably do it differently. Sorry!

**Review by justingrant — APPROVED** [review]
Great work here!

**Comment by justingrant** [comment]
@12wrigja - Actually, one more thing: I did some research about bundling dependencies, and I think CJS should *not* bundle its dependencies. I think that this will also solve any sourcemap paths issues for anyone using non-ESM bundles in node.

The change would be simple: 

```js
export default [
  {
    input,
    external,
    output: [
      outputEntry(pkg.module, 'es'),
      outputEntry(pkg.main, 'cjs')
    ],
    plugins
  },
  {
    input,
    output: [
      outputEntry(pkg.browser, 'umd'),
      outputEntry(pkg.module.replace('.esm', '.browser.esm'), 'es')
    ],
    plugins
  }
];
```

What do you think?

**Comment by 12wrigja** [comment]
@justingrant Done. Pity about the linter formatting though.

**Comment by 12wrigja** [comment]
I've rebased this on top of main and squashed in the adjustments to the rollup config. Is there anything left to resolve, or can we merge this now?

**Pull request merged**
